### PR TITLE
abci: skip repeated tx sig verification in proposal processing

### PR DESCRIPTION
**Pull request by jchappelow**
Profiling (`http/pprof`) under heavy transaction load revealed that there was major CPU utilization in `ProcessProposal` -> `validateProposalTransactions` at `ident.VerifyTransaction`.  This PR allows to skip that if the transaction (identified by hash) has already gone through this check in `CheckTx`, which would happen either via p2p gossip or rpc broadcast of the transaction.

WIP since there's a `txCache` map with a slightly different purpose that I may try to remove.

With this change, in a network with multiple validators that have to process another validator's block proposal, transaction throughput is much higher on account of much faster block proposal processing (can vanish almost 2 seconds).

### CODE REVIEW

**Comment by jchappelow** [comment]
Here's the hotspot that this PR is addressing (for cgo build):

image

for non-cgo build, same hotspot in our app, but different:

image

File: cmd/kwild/server/build.go [hunk 1]
```diff
@@ -622,6 +622,11 @@ func buildStatesyncer(d *coreDependencies) *statesync.StateSyncer {
 			continue
 		}
 
+		if res.Header.Height == 0 {
+			d.log.Warnf("zero height from provider %v", p)
+			continue
+		}
```
Comment (jchappelow) [inline cmd/kwild/server/build.go:628]: Somewhat unrelated change here.  @charithabandi and I could not understand why sometimes a bad RPC provider would get past this build step without triggering the `failBuild` below.  My guess was that perhaps it was getting a nonsense response, but no error. :man_shrugging: wish I had checked the logs when it happened.

File: internal/abci/abci.go [hunk 2]
```diff
@@ -66,21 +64,30 @@ func NewAbciApp(ctx context.Context, cfg *AbciConfig, snapshotter SnapshotModule
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to get validators: %w", err)
+	}
+	for _, val := range validators {
+		addr, err := pubkeyToAddr(val.PubKey)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert pubkey to address: %w", err)
+		}
+
+		app.validatorAddressToPubKey[addr] = val.PubKey
```
Comment (jchappelow) [inline internal/abci/abci.go:88]: Also somewhat unrelated other than being in abci, but this was moved from the `Info` method, which is not part of cometbft's consensus connection methods, making it not thread safe.  The map isn't guarded, so it should only be touched in constructor or the consensus connection methods that are used sequentially.

Honestly I don't know why this was done in `Info`, but my guess is that we were using it like a constructor where we only had DB there.  But now we've got DB from the outset, so this is an easy fix now.

File: internal/abci/abci.go [hunk 3]
```diff
@@ -478,10 +501,9 @@ func (a *AbciApp) FinalizeBlock(ctx context.Context, req *abciTypes.RequestFinal
 		res.TxResults = append(res.TxResults, abciRes)
 
 		// Remove the transaction from the cache as it has been included in a block
-		hash := cmtTypes.Tx(tx).Hash()
```
Comment (jchappelow) [inline internal/abci/abci.go:481]: Here's another optimization in `FinalizeBlock`: it hashed each transaction twice.

File: internal/abci/abci.go [hunk 4]
```diff
@@ -1210,8 +1219,12 @@ func (a *AbciApp) validateProposalTransactions(ctx context.Context, txns [][]byt
 
 			// This block proposal may include transactions that did not pass
 			// through our mempool, so we have to verify all signatures.
-			if err = ident.VerifyTransaction(tx); err != nil {
-				return fmt.Errorf("transaction signature verification failed: %w", err)
+			if !a.TxSigVerified(txO.hash) {
+				if err = ident.VerifyTransaction(tx); err != nil {
```
Comment (jchappelow) [inline internal/abci/abci.go:1223]: This is the headline change in the PR.  Validators processing block proposals would do repeated and expensive signature verification.

**Comment by jchappelow** [comment]
@charithabandi We've been running the tests with this one.  Look OK to merge?

**Review by charithabandi — APPROVED** [review]
lgtm!
